File: app/src/main/java/com/example/jacob/myapplication/Logic/ConversationData.java
package com.example.jacob.myapplication.Logic;

import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.data.PieData;
import com.github.mikephil.charting.data.PieDataSet;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.SortedMap;
import java.util.TreeMap;

/**
 * Created by jacob on 20/11/2015.
 */
public class ConversationData implements IConversationData {
    SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy");
    private Map<String, Integer> participants = new HashMap<String, Integer>();
    private Map<String, Integer> participantsWords = new HashMap<String, Integer>();
    //private List<String> messages = new ArrayList<String>();
    private SortedMap<Date,Integer> days = new TreeMap<Date, Integer>();
    private SortedMap<Date,Integer> totalDays = new TreeMap<Date, Integer>();
    private Map<String, Integer> months = new HashMap<String, Integer>();
    private Map<String, Integer> timeofDay = new HashMap<String, Integer>();
    private String conversationName;

    public ConversationData(String conversationName) {
        this.conversationName = conversationName;
    }

    public void addData(String participant, String message, Date date, String time){
        Integer numMess = participants.get(participant);
        participants.put(participant,(numMess == null) ? 1: numMess + 1);

        Integer numDay = days.get(date);
        days.put(date,(numDay == null) ? 1: numDay + 1);

        /* -1 added for the first space at the start */
        Integer wordsAcc = participantsWords.get(participant);
        int words = message.split(" ").length - 1;
        participantsWords.put(participant,(wordsAcc == null) ? words : wordsAcc + words);

        /*if (!message.equals(" <Archivo omitido>"))
            messages.add(message);*/

        Integer timeCount = timeofDay.get(time);
        timeofDay.put(time, (timeCount == null) ? 1 : timeCount + 1);
    }

    public void addData(AnalyzedLine a){
        addData(a.getParticipant(), a.getMessage(), a.getDate(), a.getHours());
    }

    public void createMonthsData() {
        SimpleDateFormat sdf = new SimpleDateFormat("MM-yyyy");
        Integer numMonth;
        for (Date iterator : days.keySet()){
            numMonth = months.get(sdf.format(iterator));
            months.put(sdf.format(iterator),(numMonth == null) ? days.get(iterator) : numMonth + days.get(iterator));
        }
    }

    public void createTotalDaysData() {
        Date iterator = new Date(days.firstKey().getTime());
        Date last = new Date(System.currentTimeMillis());
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(iterator);
        for (; iterator.before(last); calendar.add(Calendar.DAY_OF_YEAR, 1)) {
            iterator = calendar.getTime();
            totalDays.put(iterator, (days.containsKey(iterator)) ? days.get(iterator) : 0);
        }
    }

    @Override
    public List<String> getParticipants(){
        return new ArrayList<String>(participants.keySet());
    }

    @Override
    public Date getMostTalkedDay(){
        int msg = 0;
        Date date = null;
        for (Date iterator : days.keySet()){
            if (days.get(iterator) > msg) {
                msg = days.get(iterator);
                date = iterator;
            }
        }
        return date;
    }

    @Override
    public String getMostTalkedMonth() {
        String mostalkedMonth = "";
        int msg = 0;

        for (String mt : months.keySet()) {
            if (months.get(mt) > msg) {
                msg = months.get(mt);
                mostalkedMonth = mt;
            }
        }
        return mostalkedMonth;
    }

    @Override
    public String getConversationName() {
        return conversationName;
    }

    @Override
    public int getDayData(Date date) {
        return days.get(date);
    }

    @Override
    public int getMonthData(String month) {
        return months.get(month);
    }

    @Override
    public int getTotalMessages(){
        int total = 0;
        for (String iterator : participants.keySet())
            total = total + participants.get(iterator);
        return total;
    }

    @Override
    public int getTotalDaysTalked() {
        return totalDays.size();
    }

    @Override
    public int getParticipantCount(String pt) {
        return participants.get(pt);
    }

    @Override
    public float getParticipantShare(String pt) {
        float avg;
        int tot = getTotalMessages();
        avg =(float) (participants.get(pt)*100.0)/tot;
        return avg;
    }

    @Override
    public float getWordsAvg(String pt) {
        return (float) participantsWords.get(pt) / participants.get(pt);
    }

    @Override
    public float getDailyAvg() {
        return (float) getTotalMessages() / days.size();
    }

    @Override
    public float getRealDailyAvg() {
        return (float) getTotalMessages() / totalDays.size();
    }

   /* public List<String> getMessagesContent(){
        return messages;
    }*/

    public LineData getChartDaysData(){


        int i = 0;
        ArrayList<Entry> vals = new ArrayList<Entry>();
        ArrayList<String> axis = new ArrayList<String>();

        for (Date date : days.keySet()) {
            Entry a = new Entry(days.get(date),i);
            vals.add(a);
            axis.add(sdf.format(date));
            i++;
        }
        LineDataSet dt = new LineDataSet(vals, "Messages");
        dt.setAxisDependency(YAxis.AxisDependency.LEFT);
        return new LineData(axis,dt);
    }

    public LineData getAllDaysChartData(){
        int i = 0;
        ArrayList<Entry> vals = new ArrayList<Entry>();
        ArrayList<String> axis = new ArrayList<String>();

        for (Date date : totalDays.keySet()) {
            Entry a = new Entry(totalDays.get(date),i);
            vals.add(a);
            axis.add(sdf.format(date));
            i++;
        }
        LineDataSet dt = new LineDataSet(vals, "Messages");
        dt.setAxisDependency(YAxis.AxisDependency.LEFT);
        return new LineData(axis,dt);
    }
    /*
    public List<PieChart.Data> getChartParticipantsData(){
        List<PieChart.Data> data = new ArrayList<>();
        for (String s : participants.keySet()) {
            data.add(new PieChart.Data(s, getParticipantShare(s)));
        }
        return data;
    }*/

    public PieData getTimeChartData(){
        int i = 0;
        ArrayList<Entry> vals = new ArrayList<Entry>();
        ArrayList<String> axis = new ArrayList<String>();
        for (String s : timeofDay.keySet()) {
            Entry a = new Entry(timeofDay.get(s),i);
            vals.add(a);
            axis.add(s);
            i++;
        }
        PieDataSet dt = new PieDataSet(vals, "Time of the day");
        dt.setAxisDependency(YAxis.AxisDependency.LEFT);
        return new PieData(axis,dt);
    }
    /*
    public Dataset getDaysDataSet() {
        DefaultCategoryDataset dataset = new DefaultCategoryDataset();
        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");
        for (Date iterator : days.keySet()) {
            dataset.addValue(days.get(iterator), "Mensajes por dia", sdf.format(iterator));
        }
        return dataset;
    }

    public Dataset getTotalDaysDataSet() {
        DefaultCategoryDataset dataset = new DefaultCategoryDataset();
        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");

        for (Date iterator : totalDays.keySet()) {
            dataset.addValue(totalDays.get(iterator),"Mensajes por dia",sdf.format(iterator));
        }
        return dataset;
    }*/
}

File: app/src/main/java/com/example/jacob/myapplication/logic/ConversationData.kt
package com.example.jacob.myapplication.logic

import com.example.jacob.myapplication.analysis.Person
import com.example.jacob.myapplication.analysis.WordAnalyzer
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.*
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by jacob on 20/11/2015.s
 */
class ConversationData(private val conversationName: String) : IConversationData {
    internal var sdf = SimpleDateFormat("dd/MM/yyyy")
    private val participants = HashMap<String, Int>()
    private val participantsWords = HashMap<String, Int>()
    private val messages = HashMap<String, List<String>>()
    private val days = TreeMap<Date, Int>()
    private val totalDays = TreeMap<Date, Int>()
    private val months = HashMap<String, Int>()
    private val timeofDay = HashMap<String, Int>()
    private val personData: MutableList<Person> = ArrayList()

    fun addData(participant: String, message: String, date: Date, time: String) {
        val numMess = participants[participant]
        participants.put(participant, if (numMess == null) 1 else numMess + 1)

        val numDay = days[date]
        days.put(date, if (numDay == null) 1 else numDay + 1)

        /* -1 added for the first space at the start */
        val wordsAcc = participantsWords[participant]
        val words = message.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().size - 1
        participantsWords.put(participant, if (wordsAcc == null) words else wordsAcc + words)

        /*if (!message.equals(" <Archivo omitido>"))
            messages.add(message);*/

        val timeCount = timeofDay[time]
        timeofDay.put(time, if (timeCount == null) 1 else timeCount + 1)

        var msgsList: ArrayList<String>? = messages[participant] as ArrayList<String>?
        if (msgsList == null)
            msgsList = ArrayList<String>()
        msgsList.add(message)
        messages.put(participant, msgsList)
    }

    fun addData(a: AnalyzedLine?) {
        if (a != null) {
            addData(a.participant, a.message, a.date, a.hours)
        }

    }

    fun createMonthsData() {
        val sdf = SimpleDateFormat("MM-yyyy")
        for (iterator in days.keys) {
            months.merge(sdf.format(iterator), days.getValue(iterator), { currentValue, daysValue -> currentValue + daysValue })
        }
    }

    fun createTotalDaysData() {
        var iterator = Date(days.firstKey().time)
        val last = Date(System.currentTimeMillis())
        val calendar = Calendar.getInstance()
        calendar.time = iterator
        while (iterator.before(last)) {
            iterator = calendar.time
            totalDays.put(iterator, if (days.containsKey(iterator)) days.getValue(iterator) else 0)
            calendar.add(Calendar.DAY_OF_YEAR, 1)
        }
    }

    fun createPeopleData() {
        val wa = WordAnalyzer()
        participants.keys.map { personData.add(wa.analyze(it, messages.getValue(it))) }
    }

    fun getPersonData(i: Int): Person {
        return personData[i]
    }

    override fun getParticipants(): List<String> {
        return ArrayList(participants.keys)
    }

    override fun getMostTalkedDay(): Date {
        return days.maxBy { it.value }?.key ?: Date(0)
    }

    override fun getMostTalkedMonth(): String {
        return months.maxBy { it.value }!!.key
    }

    override fun getConversationName(): String {
        return conversationName
    }

    override fun getDayData(date: Date): Int {
        return days.getValue(date)
    }

    override fun getMonthData(month: String): Int {
        return months.getValue(month)
    }

    override fun getTotalMessages(): Int {
        return participants.map { it.value }.sum()
    }

    override fun getTotalDaysTalked(): Int {
        return totalDays.size
    }

    override fun getParticipantCount(pt: String): Int {
        return participants.getValue(pt)
    }

    override fun getParticipantShare(pt: String): Float {
        return (participants.getValue(pt) * 100.0).toFloat() / totalMessages
    }

    override fun getWordsAvg(pt: String): Float {
        return participantsWords.getValue(pt) / participants.getValue(pt).toFloat()
    }

    override fun getDailyAvg(): Float {
        return totalMessages.toFloat() / days.size
    }

    override fun getRealDailyAvg(): Float {
        return totalMessages.toFloat() / totalDays.size
    }

    val chartDaysData: Pair<LineData, List<String>>
        get() {
            var i = 0
            val vals = ArrayList<Entry>()
            val axis = ArrayList<String>()


            days.map {
                vals.add(Entry(i++.toFloat(), it.value.toFloat()))
                axis.add(sdf.format(it.key))
            }

            val dt = LineDataSet(vals, "Messages")
            //dt.axisDependency = YAxis.AxisDependency.LEFT
            return Pair(LineData(dt), axis)
        }

    val allDaysChartData: Pair<LineData, List<String>>
        get() {
            var i = 0
            val vals = ArrayList<Entry>()
            val axis = ArrayList<String>()

            totalDays.map {
                vals.add(Entry((i++).toFloat(), it.value.toFloat()))
                axis.add(sdf.format(it.key))
            }
            val dt = LineDataSet(vals, "Messages")

            dt.axisDependency = YAxis.AxisDependency.LEFT
            return Pair(LineData(dt), axis)
        }
    /*
    public List<PieChart.Data> getChartParticipantsData(){
        List<PieChart.Data> data = new ArrayList<>();
        for (String s : participants.keySet()) {
            data.add(new PieChart.Data(s, getParticipantShare(s)));
        }
        return data;
    }*/

    val timeChartData: PieData
        get() {
            val vals = ArrayList<PieEntry>()

            timeofDay.map {
                vals.add(PieEntry(it.value.toFloat(), it.key))
            }

            val dt = PieDataSet(vals, "Time of the day")
            dt.axisDependency = YAxis.AxisDependency.LEFT
            return PieData(dt)
        }
    /*
    public Dataset getDaysDataSet() {
        DefaultCategoryDataset dataset = new DefaultCategoryDataset();
        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");
        for (Date iterator : days.keySet()) {
            dataset.addValue(days.get(iterator), "Mensajes por dia", sdf.format(iterator));
        }
        return dataset;
    }

    public Dataset getTotalDaysDataSet() {
        DefaultCategoryDataset dataset = new DefaultCategoryDataset();
        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy");

        for (Date iterator : totalDays.keySet()) {
            dataset.addValue(totalDays.get(iterator),"Mensajes por dia",sdf.format(iterator));
        }
        return dataset;
    }*/
}
